client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java

```diff
@@ -170,35 +175,59 @@ public List<ShuffleBlockInfo> addRecord(int partitionId, Object key, Object valu
     if (serializedDataLength == 0) {
       return null;
     }
-    List<ShuffleBlockInfo> result = Lists.newArrayList();
+
+    List<ShuffleBlockInfo> sentBlocks = Lists.newArrayList();
+    requestMemoryAndInsert(partitionId, serializedData, serializedDataLength, sentBlocks);
+    shuffleWriteMetrics.incRecordsWritten(1L);
+
+    if (usedBytes.get() - inSendListBytes.get() > spillSize || spillTriggered.compareAndSet(true, false)) {
```

Comment:
It's weird synchroinization.

Reply:
Nice catch. It's bug

Reply:
You design too complex synchronization mechnism. It will be buggy.

Reply:
I think it's simple that all sync operation has been changed to one thread to reduce cost of sync.

The key point is to send signal to the thread of adding records when spilling in another thread.

Feel free to propose your idea

Reply:
Why don't we use lock directly?

Reply:
Bring too much cost. Every record with one lock is unreasonable.

Reply:
> Bring too much cost. Every record with one lock is unreasonable.

If we don't have lock conflicts, it won't bring too much cost. It is about several nano seconds.

Reply:
Optimistic lock will. But I still think this is not complex. @advancedxy If you have time, could you help check this?

Reply:
> if `SpillTrigger` is syncrhonization value, why do we put the variable after `||`? It's strange.

Done.

Reply:
> Optimistic lock will. But I still think this is not complex. @advancedxy If you have time, could you help check this?

In my opinion, there is a bug, it reflects that the mechinsm is too complex.